### PR TITLE
09-coap: add CoAP release specs

### LINKED ISSUE #73: RFC: CoAP release specs
With https://github.com/RIOT-OS/RIOT/pull/7428 I was wondering if we could integrate CoAP into our release specifications (or actually astonished that we still don't include it). It is a very important protocol for the IoT and should thus be rigorously tested (IMHO tests with native should be enough though). We could use the plugtests for CoAP and CoAP-RD that were done at various IETF-related events as a template for that.

### PR DESCRIPTION
**Pull request by kb2ma**
### Contribution description
Basic coverage for CoAP messaging as described in #73. The goal for this PR is to include non-confirmable and confirmable messaging, block extension, and observe extension. Includes some tests for both gcoap and nanocoap. Includes test of RIOT Resource Directory client implementation (CORD).

Initial commit probably premature, but I wanted to get the message out that I'm working on it for the 2018.10 release. :-)

### Issues/PRs references
Resolves #73

### CODE REVIEW

**Comment by kb2ma** [comment]
Making good progress; one more task to go. I did just find a bug in nanocoap that probably should be fixed for the release, so I'll create a PR for that before continuing with the test spec.

**Comment by cladmi** [comment]
Good to have dedicated tests using `coap` for the release.

A remark that shows a flaw in my previous naming, maybe my task 09 should be named 99 as I wanted it to be an optional bonus task. This way your task would be a more important task directly after the others. I can rename it.

**Comment by jia200x** [comment]
thanks @kb2ma !!

**Comment by kb2ma** [comment]
Sure thing @jia200x, glad to get this in. It gave me an excuse to work with aiocoap, which is worthwhile.

@cladmi, I'll follow your lead on the naming for the directory. If 09-compile... goes away, I'll rename 10-coap to take its place.

**Comment by cladmi** [comment]
It went away :)

**Comment by kb2ma** [comment]
Added Observe task, renamed to 09-coap. Squashed and ready for review. The content should be usable by others.

Also, from my perspective gcoap and nanocoap work fine in 2018.10-branch with the fix in [RIOT #10298](https://github.com/RIOT-OS/RIOT/pull/10298).

File: 09-coap/coap.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+```
+RD endpoint event: successfully updated client registration
+```
+
+
+Task #02 - Confirmable retries
+==============================
+### Description
+
+gcoap sends a CON GET requst to an aiocoap server. The test runner  delays startup of aiocoap to trigger gcoap to resend the message.
```
Comment (aabadie) [inline 09-coap/coap.md]: s/requst/request/ + 2 spaces between `runner` and `delays`

File: 09-coap/coap.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+```
+gcoap: timeout for msg ID 50754
+```
+
+
+Task #03 - Block1
+=================
+### Description
+
+aiocoap slices a payload into a sequence of Block1 encoded POSTs to nanocoap. nanocoap computes and returns the SHA-256 message digest on the payload.
```
Comment (aabadie) [inline 09-coap/coap.md]: s/on the payload/of the payload/ ?

**Comment by kb2ma** [comment]
Pushed the wrong button. :-)

**Comment by kb2ma** [comment]
And of course I'm happy to run these tests for the release.

**Comment by kb2ma** [comment]
Squashed, looks good to me.

**Comment by aabadie** [comment]
Thanks, let's merge!

**Pull request merged**